crates/ty_python_semantic/src/types.rs

```diff
@@ -6958,6 +6733,55 @@ impl<'db> ConstructorCallError<'db> {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(crate) enum TypeRelation {
+    Subtyping,
+    Assignability,
+}
+
+impl TypeRelation {
+    /// Non-fully-static types do not participate in subtyping, only assignability,
+    /// so the subtyping relation does not even apply to them.
+    ///
+    /// Type `A` can only be a subtype of type `B` if the set of possible runtime objects
+    /// that `A` represents is a subset of the set of possible runtime objects that `B` represents.
+    /// But the set of objects described by a non-fully-static type is (either partially or wholly) unknown,
+    /// so the question is simply unanswerable for non-fully-static types.
+    ///
+    /// However, the assignability relation applies to all types, even non-fully-static ones.
+    fn applies_to<'db, I>(self, db: &'db dyn Db, type_1: I, type_2: I) -> bool
+    where
+        I: Into<Type<'db>>,
+    {
```

Comment:
nit: this isn't a large method so feel free to ignore but I was wondering whether we should enforce `type_1: Type<'db>` instead of the `Into` trait and make the changes at the call side `.applies_to(db, self.into(), other.into())`. This is mainly to avoid monomorphization of this method to the different combinations of types used with this method.